dipy/segment/clustering.py

```diff
@@ -0,0 +1,372 @@
+    """ Provides functionalities to interact with clustering outputs.
+
+    Useful container to create, remove, retrieve and filter clusters.
+    It allows also to retrieve easely the centroid of every clusters.
+    If `refdata` is given, elements will be returned instead of their
+    index when using `ClusterCentroid` objects.
+
+    Parameters
+    ----------
+    refdata : list
+        Actual elements that clustered indices refer to.
+    """
+    @property
+    def centroids(self):
+        return [cluster.centroid for cluster in self.clusters]
+
+
+class Clustering:
+    def cluster(self, data, ordering=None):
+        """ Clusters `data`.
+
+        Subclasses will perform their clustering algorithm here.
+
+        Parameters
+        ----------
+        data : list of N-dimensional array
+            Each array represents a data point.
+        ordering : iterable of indices, optional
+            Specifies the order in which data points will be clustered.
+
+        Returns
+        -------
+        `ClusterMap` object
+            Result of the clustering.
+        """
+        raise NotImplementedError("Subclass has to define method 'cluster(data, ordering)'!")
+
+
+class QuickBundles(Clustering):
+    r""" Clusters streamlines using QuickBundles algorithm.
+
+    Given a list of streamlines, the QuickBundles algorithm sequentially
+    assigns each streamline to its closest bundle in $\mathcal{O}(Nk)$ where
+    $N$ is the number of streamlines and $k$ is the final number of bundles.
+    If for a given streamline its closest bundle is farther than `threshold`,
+    a new bundle is created and the streamline is assigned to it except if the
+    number of bundles has already exceeded `max_nb_clusters`.
+
+    Parameters
+    ----------
+    threshold : float
+        The maximum distance from a bundle for a streamline to be still
+        considered as part of it.
+    metric : str or `Metric` object
+        The distance metric to use when comparing two streamlines.
+    max_nb_clusters : int
+        Limits the creation of bundles.
+
+    References
+    ----------
+    .. [Garyfallidis12] Garyfallidis E. et al., QuickBundles a method for
+                        tractography simplification, Frontiers in Neuroscience,
+                        vol 6, no 175, 2012.
+    """
+    def __init__(self, threshold, metric="mdf", max_nb_clusters=np.iinfo('i4').max):
+        self.threshold = threshold
+        self.max_nb_clusters = max_nb_clusters
+
+        if isinstance(metric, Metric):
+            self.metric = metric
+        elif metric.lower() == "mdf":
+            self.metric = AveragePointwiseEuclideanMetric()
+
+    def cluster(self, streamlines, ordering=None, refdata=None):
+        """ Clusters `streamlines` into bundles.
+
+        Performs quickbundles algorithm using predefined metric and threshold.
+
+        Parameters
+        ----------
+        streamlines : list (or generator) of 2D array
```

Comment:
Actually this is a typo. Presently, `streamlines` needs to be a list of 2D array. I have another version where it could accept either a list or a generator of 2D array though.

However, this is not related with the sequential clustering if by this you mean clustering one streamline at the time while updating a ClusterMap?. I did not find an efficient way to do it right now.

Hmm, after thinking about it, it would be really nice to accept generator, especially with the new NiBabel streamlines API where you can specify to not load all streamlines in memory. I should probably wait for this PR to be accepted before adding that option though.